<!DOCTYPE html >
<html>
        <head>
          <meta http-equiv="X-UA-Compatible" content="IE=edge" />
          <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" />
          <title>Smile - Statistical Machine Intelligence and Learning Engine  - smile.read</title>
          <meta name="description" content="Smile - Statistical Machine Intelligence and Learning Engine - smile.read" />
          <meta name="keywords" content="Smile Statistical Machine Intelligence and Learning Engine smile.read" />
          <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
          
      <link href="../lib/index.css" media="screen" type="text/css" rel="stylesheet" />
      <link href="../lib/template.css" media="screen" type="text/css" rel="stylesheet" />
      <link href="../lib/diagrams.css" media="screen" type="text/css" rel="stylesheet" id="diagrams-css" />
      <script type="text/javascript" src="../lib/jquery.js"></script>
      <script type="text/javascript" src="../lib/jquery.panzoom.min.js"></script>
      <script type="text/javascript" src="../lib/jquery.mousewheel.min.js"></script>
      <script type="text/javascript" src="../lib/index.js"></script>
      <script type="text/javascript" src="../index.js"></script>
      <script type="text/javascript" src="../lib/scheduler.js"></script>
      <script type="text/javascript" src="../lib/template.js"></script>
      <script type="text/javascript" src="../lib/tools.tooltip.js"></script>
      
      <script type="text/javascript">
        /* this variable can be used by the JS to determine the path to the root document */
        var toRoot = '../';
      </script>
    
        </head>
        <body>
      <div id="search">
        <span id="doc-title">Smile - Statistical Machine Intelligence and Learning Engine<span id="doc-version"></span></span>
        <span class="close-results"><span class="left">&lt;</span> Back</span>
        <div id="textfilter">
          <span class="input">
            <input autocapitalize="none" placeholder="Search" id="index-input" type="text" accesskey="/" />
            <i class="clear material-icons"></i>
            <i id="search-icon" class="material-icons"></i>
          </span>
        </div>
    </div>
      <div id="search-results">
        <div id="search-progress">
          <div id="progress-fill"></div>
        </div>
        <div id="results-content">
          <div id="entity-results"></div>
          <div id="member-results"></div>
        </div>
      </div>
      <div id="content-scroll-container" style="-webkit-overflow-scrolling: touch;">
        <div id="content-container" style="-webkit-overflow-scrolling: touch;">
          <div id="subpackage-spacer">
            <div id="packages">
              <h1>Packages</h1>
              <ul>
                <li name="_root_.root" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="_root_"></a><a id="root:_root_"></a>
      <span class="permalink">
      <a href="index.html#_root_" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="High level Smile operators in Scala." href="../index.html">
              <span class="name">root</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">High level Smile operators in Scala.</p><div class="fullcomment"><div class="comment cmt"><p>High level Smile operators in Scala.</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../index.html" class="extype" name="_root_">root</a></dd></dl></div>
    </li><li name="_root_.smile" visbl="pub" class="indented1 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="smile"></a><a id="smile:smile"></a>
      <span class="permalink">
      <a href="index.html#smile" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="" href="index.html">
              <span class="name">smile</span>
            </a>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../index.html" class="extype" name="_root_">root</a></dd></dl></div>
    </li><li name="smile.association" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="association"></a><a id="association:association"></a>
      <span class="permalink">
      <a href="../smile/index.html#association" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Frequent item set mining and association rule mining." href="association/index.html">
              <span class="name">association</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Frequent item set mining and association rule mining.</p><div class="fullcomment"><div class="comment cmt"><p>Frequent item set mining and association rule mining.
Association rule learning is a popular and well researched method for
discovering interesting relations between variables in large databases.
Let I = {i<sub>1</sub>, i<sub>2</sub>,..., i<sub>n</sub>} be a set of n
binary attributes called items. Let D = {t<sub>1</sub>, t<sub>2</sub>,..., t<sub>m</sub>}
be a set of transactions called the database. Each transaction in D has a
unique transaction ID and contains a subset of the items in I.
An association rule is defined as an implication of the form X &rArr; Y
where X, Y &sube; I and X &cap; Y = &Oslash;. The item sets X and Y are called
antecedent (left-hand-side or LHS) and consequent (right-hand-side or RHS)
of the rule, respectively. The support supp(X) of an item set X is defined as
the proportion of transactions in the database which contain the item set.
Note that the support of an association rule X &rArr; Y is supp(X &cup; Y).
The confidence of a rule is defined conf(X &rArr; Y) = supp(X &cup; Y) / supp(X).
Confidence can be interpreted as an estimate of the probability P(Y | X),
the probability of finding the RHS of the rule in transactions under the
condition that these transactions also contain the LHS.</p><p>For example, the rule {onions, potatoes} &rArr; {burger} found in the sales
data of a supermarket would indicate that if a customer buys onions and
potatoes together, he or she is likely to also buy burger. Such information
can be used as the basis for decisions about marketing activities such as
promotional pricing or product placements.</p><p>Association rules are usually required to satisfy a user-specified minimum
support and a user-specified minimum confidence at the same time. Association
rule generation is usually split up into two separate steps:</p><ul><li>First, minimum support is applied to find all frequent item sets
   in a database (i.e. frequent item set mining).</li><li>Second, these frequent item sets and the minimum confidence constraint
   are used to form rules.</li></ul><p>Finding all frequent item sets in a database is difficult since it involves
searching all possible item sets (item combinations). The set of possible
item sets is the power set over I (the set of items) and has size 2<sup>n</sup> - 1
(excluding the empty set which is not a valid item set). Although the size
of the power set grows exponentially in the number of items n in I, efficient
search is possible using the downward-closure property of support
(also called anti-monotonicity) which guarantees that for a frequent item set
also all its subsets are frequent and thus for an infrequent item set, all
its supersets must be infrequent.</p><p>In practice, we may only consider the frequent item set that has the maximum
number of items bypassing all the sub item sets. An item set is maximal
frequent if none of its immediate supersets is frequent.</p><p>For a maximal frequent item set, even though we know that all the sub item
sets are frequent, we don't know the actual support of those sub item sets,
which are very important to find the association rules within the item sets.
If the final goal is association rule mining, we would like to discover
closed frequent item sets. An item set is closed if none of its immediate
supersets has the same support as the item set.</p><p>Some well known algorithms of frequent item set mining are Apriori,
Eclat and FP-Growth. Apriori is the best-known algorithm to mine association
rules. It uses a breadth-first search strategy to counting the support of
item sets and uses a candidate generation function which exploits the downward
closure property of support. Eclat is a depth-first search algorithm using
set intersection.</p><p>FP-growth (frequent pattern growth) uses an extended prefix-tree (FP-tree)
structure to store the database in a compressed form. FP-growth adopts a
divide-and-conquer approach to decompose both the mining tasks and the
databases. It uses a pattern fragment growth method to avoid the costly
process of candidate generation and testing used by Apriori.</p><h6>References:</h6><ul><li>R. Agrawal, T. Imielinski and A. Swami. Mining Association Rules Between Sets of Items in Large Databases, SIGMOD, 207-216, 1993.</li><li>Rakesh Agrawal and Ramakrishnan Srikant. Fast algorithms for mining association rules in large databases. VLDB, 487-499, 1994.</li><li>Mohammed J. Zaki. Scalable algorithms for association mining. IEEE Transactions on Knowledge and Data Engineering, 12(3):372-390, 2000.</li><li>Jiawei Han, Jian Pei, Yiwen Yin, and Runying Mao. Mining frequent patterns without candidate generation. Data Mining and Knowledge Discovery 8:53-87, 2004.
</li></ul></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.classification" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="classification"></a><a id="classification:classification"></a>
      <span class="permalink">
      <a href="../smile/index.html#classification" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Classification algorithms." href="classification/index.html">
              <span class="name">classification</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Classification algorithms.</p><div class="fullcomment"><div class="comment cmt"><p>Classification algorithms. In machine learning and pattern recognition,
classification refers to an algorithmic procedure for assigning a given
input object into one of a given number of categories. The input
object is formally termed an instance, and the categories are termed classes.</p><p>The instance is usually described by a vector of features, which together
constitute a description of all known characteristics of the instance.
Typically, features are either categorical (also known as nominal, i.e.
consisting of one of a set of unordered items, such as a gender of &quot;male&quot;
or &quot;female&quot;, or a blood type of &quot;A&quot;, &quot;B&quot;, &quot;AB&quot; or &quot;O&quot;), ordinal (consisting
of one of a set of ordered items, e.g. &quot;large&quot;, &quot;medium&quot; or &quot;small&quot;),
integer-valued (e.g. a count of the number of occurrences of a particular
word in an email) or real-valued (e.g. a measurement of blood pressure).</p><p>Classification normally refers to a supervised procedure, i.e. a procedure
that produces an inferred function to predict the output value of new
instances based on a training set of pairs consisting of an input object
and a desired output value. The inferred function is called a classifier
if the output is discrete or a regression function if the output is
continuous.</p><p>The inferred function should predict the correct output value for any valid
input object. This requires the learning algorithm to generalize from the
training data to unseen situations in a &quot;reasonable&quot; way.</p><p>A wide range of supervised learning algorithms is available, each with
its strengths and weaknesses. There is no single learning algorithm that
works best on all supervised learning problems. The most widely used
learning algorithms are AdaBoost and gradient boosting, support vector
machines, linear regression, linear discriminant analysis, logistic
regression, naive Bayes, decision trees, k-nearest neighbor algorithm,
and neural networks (multilayer perceptron).</p><p>If the feature vectors include features of many different kinds (discrete,
discrete ordered, counts, continuous values), some algorithms cannot be
easily applied. Many algorithms, including linear regression, logistic
regression, neural networks, and nearest neighbor methods, require that
the input features be numerical and scaled to similar ranges (e.g., to
the [-1,1] interval). Methods that employ a distance function, such as
nearest neighbor methods and support vector machines with Gaussian kernels,
are particularly sensitive to this. An advantage of decision trees (and
boosting algorithms based on decision trees) is that they easily handle
heterogeneous data.</p><p>If the input features contain redundant information (e.g., highly correlated
features), some learning algorithms (e.g., linear regression, logistic
regression, and distance based methods) will perform poorly because of
numerical instabilities. These problems can often be solved by imposing
some form of regularization.</p><p>If each of the features makes an independent contribution to the output,
then algorithms based on linear functions (e.g., linear regression,
logistic regression, linear support vector machines, naive Bayes) generally
perform well. However, if there are complex interactions among features,
then algorithms such as nonlinear support vector machines, decision trees
and neural networks work better. Linear methods can also be applied, but
the engineer must manually specify the interactions when using them.</p><p>There are several major issues to consider in supervised learning:</p><ul><li><b>Features:</b>
The accuracy of the inferred function depends strongly on how the input
object is represented. Typically, the input object is transformed into
a feature vector, which contains a number of features that are descriptive
of the object. The number of features should not be too large, because of
the curse of dimensionality; but should contain enough information to
accurately predict the output.
There are many algorithms for feature selection that seek to identify
the relevant features and discard the irrelevant ones. More generally,
dimensionality reduction may seek to map the input data into a lower
dimensional space prior to running the supervised learning algorithm.</li><li><b>Overfitting:</b>
Overfitting occurs when a statistical model describes random error
or noise instead of the underlying relationship. Overfitting generally
occurs when a model is excessively complex, such as having too many
parameters relative to the number of observations. A model which has
been overfit will generally have poor predictive performance, as it can
exaggerate minor fluctuations in the data.
The potential for overfitting depends not only on the number of parameters
and data but also the conformability of the model structure with the data
shape, and the magnitude of model error compared to the expected level
of noise or error in the data.
In order to avoid overfitting, it is necessary to use additional techniques
(e.g. cross-validation, regularization, early stopping, pruning, Bayesian
priors on parameters or model comparison), that can indicate when further
training is not resulting in better generalization. The basis of some
techniques is either (1) to explicitly penalize overly complex models,
or (2) to test the model's ability to generalize by evaluating its
performance on a set of data not used for training, which is assumed to
approximate the typical unseen data that a model will encounter.</li><li><b>Regularization:</b>
Regularization involves introducing additional information in order
to solve an ill-posed problem or to prevent over-fitting. This information
is usually of the form of a penalty for complexity, such as restrictions
for smoothness or bounds on the vector space norm.
A theoretical justification for regularization is that it attempts to impose
Occam's razor on the solution. From a Bayesian point of view, many
regularization techniques correspond to imposing certain prior distributions
on model parameters.</li><li><b>Bias-variance tradeoff:</b>
Mean squared error (MSE) can be broken down into two components:
variance and squared bias, known as the bias-variance decomposition.
Thus in order to minimize the MSE, we need to minimize both the bias and
the variance. However, this is not trivial. Therefore, there is a tradeoff
between bias and variance.
</li></ul></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.clustering" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="clustering"></a><a id="clustering:clustering"></a>
      <span class="permalink">
      <a href="../smile/index.html#clustering" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Clustering analysis." href="clustering/index.html">
              <span class="name">clustering</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Clustering analysis.</p><div class="fullcomment"><div class="comment cmt"><p>Clustering analysis. Clustering is the assignment of a set of observations
into subsets (called clusters) so that observations in the same cluster are
similar in some sense. Clustering is a method of unsupervised learning,
and a common technique for statistical data analysis used in many fields.</p><p>Hierarchical algorithms find successive clusters using previously
established clusters. These algorithms usually are either agglomerative
(&quot;bottom-up&quot;) or divisive (&quot;top-down&quot;). Agglomerative algorithms begin
with each element as a separate cluster and merge them into successively
larger clusters. Divisive algorithms begin with the whole set and proceed
to divide it into successively smaller clusters.</p><p>Partitional algorithms typically determine all clusters at once, but can
also be used as divisive algorithms in the hierarchical clustering.
Many partitional clustering algorithms require the specification of
the number of clusters to produce in the input data set, prior to
execution of the algorithm. Barring knowledge of the proper value
beforehand, the appropriate value must be determined, a problem on
its own for which a number of techniques have been developed.</p><p>Density-based clustering algorithms are devised to discover
arbitrary-shaped clusters. In this approach, a cluster is regarded as
a region in which the density of data objects exceeds a threshold.</p><p>Subspace clustering methods look for clusters that can only be seen in
a particular projection (subspace, manifold) of the data. These methods
thus can ignore irrelevant attributes. The general problem is also known
as Correlation clustering while the special case of axis-parallel subspaces
is also known as two-way clustering, co-clustering or biclustering in
bioinformatics: in these methods not only the objects are clustered but
also the features of the objects, i.e., if the data is represented in
a data matrix, the rows and columns are clustered simultaneously. They
usually do not however work with arbitrary feature combinations as in general
subspace methods.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.feature" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="feature"></a><a id="feature:feature"></a>
      <span class="permalink">
      <a href="../smile/index.html#feature" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Feature generation, normalization and selection." href="feature/index.html">
              <span class="name">feature</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Feature generation, normalization and selection.</p><div class="fullcomment"><div class="comment cmt"><p>Feature generation, normalization and selection.</p><p>Feature generation (or constructive induction) studies methods that modify
or enhance the representation of data objects. Feature generation techniques
search for new features that describe the objects better than the attributes
supplied with the training instances.</p><p>Many machine learning methods such as Neural Networks and SVM with Gaussian
kernel also require the features properly scaled/standardized. For example,
each variable is scaled into interval [0, 1] or to have mean 0 and standard
deviation 1. Although some method such as decision trees can handle nominal
variable directly, other methods generally require nominal variables converted
to multiple binary dummy variables to indicate the presence or absence of a
characteristic.</p><p>Feature selection is the technique of selecting a subset of relevant
features for building robust learning models. By removing most irrelevant
and redundant features from the data, feature selection helps improve the
performance of learning models by alleviating the effect of the curse of
dimensionality, enhancing generalization capability, speeding up learning
process, etc. More importantly, feature selection also helps researchers
to acquire better understanding about the data.</p><p>Feature selection algorithms typically fall into two categories: feature
ranking and subset selection. Feature ranking ranks the features by a
metric and eliminates all features that do not achieve an adequate score.
Subset selection searches the set of possible features for the optimal subset.
Clearly, an exhaustive search of optimal subset is impractical if large
numbers of features are available. Commonly, heuristic methods such as
genetic algorithms are employed for subset selection.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.imputation" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="imputation"></a><a id="imputation:imputation"></a>
      <span class="permalink">
      <a href="../smile/index.html#imputation" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Missing value imputation." href="imputation/index.html">
              <span class="name">imputation</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Missing value imputation.</p><div class="fullcomment"><div class="comment cmt"><p>Missing value imputation. In statistics, missing data, or missing values,
occur when no data value is stored for the variable in the current
observation. Missing data are a common occurrence and can have a
significant effect on the conclusions that can be drawn from the data.</p><p>Data are missing for many reasons. Missing data can occur because of
nonresponse: no information is provided for several items or no information
is provided for a whole unit. Some items are more sensitive for nonresponse
than others, for example items about private subjects such as income.</p><p>Dropout is a type of missingness that occurs mostly when studying
development over time. In this type of study the measurement is repeated
after a certain period of time. Missingness occurs when participants drop
out before the test ends and one or more measurements are missing.</p><p>Sometimes missing values are caused by the device failure or even by
researchers themselves. It is important to question why the data is missing,
this can help with finding a solution to the problem. If the values are
missing at random there is still information about each variable in each
unit but if the values are missing systematically the problem is more severe
because the sample cannot be representative of the population.</p><p>All of the causes for missing data fit into four classes, which are based
on the relationship between the missing data mechanism and the missing and
observed values. These classes are important to understand because the
problems caused by missing data and the solutions to these problems are
different for the four classes.</p><p>The first is Missing Completely at Random (MCAR). MCAR means that the
missing data mechanism is unrelated to the values of any variables, whether
missing or observed. Data that are missing because a researcher dropped the
test tubes or survey participants accidentally skipped questions are
likely to be MCAR. If the observed values are essentially a random sample
of the full data set, complete case analysis gives the same results as
the full data set would have. Unfortunately, most missing data are not MCAR.</p><p>At the opposite end of the spectrum is Non-Ignorable (NI). NI means that
the missing data mechanism is related to the missing values. It commonly
occurs when people do not want to reveal something very personal or
unpopular about themselves. For example, if individuals with higher incomes
are less likely to reveal them on a survey than are individuals with lower
incomes, the missing data mechanism for income is non-ignorable. Whether
income is missing or observed is related to its value. Complete case
analysis can give highly biased results for NI missing data. If
proportionally more low and moderate income individuals are left in
the sample because high income people are missing, an estimate of the
mean income will be lower than the actual population mean.</p><p>In between these two extremes are Missing at Random (MAR) and Covariate
Dependent (CD). Both of these classes require that the cause of the missing
data is unrelated to the missing values, but may be related to the observed
values of other variables. MAR means that the missing values are related to
either observed covariates or response variables, whereas CD means that the
missing values are related only to covariates. As an example of CD missing
data, missing income data may be unrelated to the actual income values, but
are related to education. Perhaps people with more education are less likely
to reveal their income than those with less education.</p><p>A key distinction is whether the mechanism is ignorable (i.e., MCAR, CD, or
MAR) or non-ignorable. There are excellent techniques for handling ignorable
missing data. Non-ignorable missing data are more challenging and require a
different approach.</p><p>If it is known that the data analysis technique which is to be used isn't
content robust, it is good to consider imputing the missing data.
Once all missing values have been imputed, the dataset can then be analyzed
using standard techniques for complete data. The analysis should ideally
take into account that there is a greater degree of uncertainty than if
the imputed values had actually been observed, however, and this generally
requires some modification of the standard complete-data analysis methods.
Many imputation techniques are available.</p><p>Imputation is not the only method available for handling missing data.
The expectation-maximization algorithm is a method for finding maximum
likelihood estimates that has been widely applied to missing data problems.
In machine learning, it is sometimes possible to train a classifier directly
over the original data without imputing it first. That was shown to yield
better performance in cases where the missing data is structurally absent,
rather than missing due to measurement noise.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.manifold" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="manifold"></a><a id="manifold:manifold"></a>
      <span class="permalink">
      <a href="../smile/index.html#manifold" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Manifold learning finds a low-dimensional basis for describing
high-dimensional data." href="manifold/index.html">
              <span class="name">manifold</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Manifold learning finds a low-dimensional basis for describing
high-dimensional data.</p><div class="fullcomment"><div class="comment cmt"><p>Manifold learning finds a low-dimensional basis for describing
high-dimensional data. Manifold learning is a popular approach to nonlinear
dimensionality reduction. Algorithms for this task are based on the idea
that the dimensionality of many data sets is only artificially high; though
each data point consists of perhaps thousands of features, it may be
described as a function of only a few underlying parameters. That is, the
data points are actually samples from a low-dimensional manifold that is
embedded in a high-dimensional space. Manifold learning algorithms attempt
to uncover these parameters in order to find a low-dimensional representation
of the data.</p><p>Some prominent approaches are locally linear embedding
(LLE), Hessian LLE, Laplacian eigenmaps, and LTSA. These techniques
construct a low-dimensional data representation using a cost function
that retains local properties of the data, and can be viewed as defining
a graph-based kernel for Kernel PCA. More recently, techniques have been
proposed that, instead of defining a fixed kernel, try to learn the kernel
using semidefinite programming. The most prominent example of such a
technique is maximum variance unfolding (MVU). The central idea of MVU
is to exactly preserve all pairwise distances between nearest neighbors
(in the inner product space), while maximizing the distances between points
that are not nearest neighbors.</p><p>An alternative approach to neighborhood preservation is through the
minimization of a cost function that measures differences between
distances in the input and output spaces. Important examples of such
techniques include classical multidimensional scaling (which is identical
to PCA), Isomap (which uses geodesic distances in the data space), diffusion
maps (which uses diffusion distances in the data space), t-SNE (which
minimizes the divergence between distributions over pairs of points),
and curvilinear component analysis.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.math" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="math"></a><a id="math:math"></a>
      <span class="permalink">
      <a href="../smile/index.html#math" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Mathematical and statistical functions." href="math/index.html">
              <span class="name">math</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Mathematical and statistical functions.</p><div class="fullcomment"><div class="comment cmt"><p>Mathematical and statistical functions.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.mds" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="mds"></a><a id="mds:mds"></a>
      <span class="permalink">
      <a href="../smile/index.html#mds" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Multidimensional scaling." href="mds/index.html">
              <span class="name">mds</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Multidimensional scaling.</p><div class="fullcomment"><div class="comment cmt"><p>Multidimensional scaling. MDS is a set of related statistical techniques
often used in information visualization for exploring similarities or
dissimilarities in data. An MDS algorithm starts with a matrix of item-item
similarities, then assigns a location to each item in N-dimensional space.
For sufficiently small N, the resulting locations may be displayed in a
graph or 3D visualization.</p><p>The major types of MDS algorithms include:</p><p><b>Classical multidimensional scaling</b>
takes an input matrix giving dissimilarities between pairs of items and
outputs a coordinate matrix whose configuration minimizes a loss function
called strain.</p><p><b>Metric multidimensional scaling</b> is
a superset of classical MDS that generalizes the optimization procedure
to a variety of loss functions and input matrices of known distances with
weights and so on. A useful loss function in this context is called stress
which is often minimized using a procedure called stress majorization.</p><p><b>Non-metric multidimensional scaling</b> finds both a non-parametric
monotonic relationship between the dissimilarities in the item-item matrix
and the Euclidean distances between items, and the location of each item in
the low-dimensional space. The relationship is typically found using isotonic
regression.</p><p><b>Generalized multidimensional scaling</b> is
an extension of metric multidimensional scaling, in which the target
space is an arbitrary smooth non-Euclidean space. In case when the
dissimilarities are distances on a surface and the target space is another
surface, GMDS allows finding the minimum-distortion embedding of one surface
into another.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.nlp" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="nlp"></a><a id="nlp:nlp"></a>
      <span class="permalink">
      <a href="../smile/index.html#nlp" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Natural language processing." href="nlp/index.html">
              <span class="name">nlp</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Natural language processing.</p><div class="fullcomment"><div class="comment cmt"><p>Natural language processing.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.plot" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="plot"></a><a id="plot:plot"></a>
      <span class="permalink">
      <a href="../smile/index.html#plot" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Data visualization." href="plot/index.html">
              <span class="name">plot</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Data visualization.</p><div class="fullcomment"><div class="comment cmt"><p>Data visualization.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.projection" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="projection"></a><a id="projection:projection"></a>
      <span class="permalink">
      <a href="../smile/index.html#projection" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Feature extraction." href="projection/index.html">
              <span class="name">projection</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Feature extraction.</p><div class="fullcomment"><div class="comment cmt"><p>Feature extraction. Feature extraction transforms the data in the
high-dimensional space to a space of fewer dimensions. The data
transformation may be linear, as in principal component analysis (PCA),
but many nonlinear dimensionality reduction techniques also exist.</p><p>The main linear technique for dimensionality reduction, principal component
analysis, performs a linear mapping of the data to a lower dimensional
space in such a way that the variance of the data in the low-dimensional
representation is maximized. In practice, the correlation matrix of the
data is constructed and the eigenvectors on this matrix are computed.
The eigenvectors that correspond to the largest eigenvalues (the principal
components) can now be used to reconstruct a large fraction of the variance
of the original data. Moreover, the first few eigenvectors can often be
interpreted in terms of the large-scale physical behavior of the system.
The original space has been reduced (with data loss, but hopefully
retaining the most important variance) to the space spanned by a few
eigenvectors.</p><p>Compared to regular batch PCA algorithm, the generalized Hebbian algorithm
is an adaptive method to find the largest k eigenvectors of the covariance
matrix, assuming that the associated eigenvalues are distinct. GHA works
with an arbitrarily large sample size and the storage requirement is modest.
Another attractive feature is that, in a nonstationary environment, it
has an inherent ability to track gradual changes in the optimal solution
in an inexpensive way.</p><p>Random projection is a promising linear dimensionality reduction technique
for learning mixtures of Gaussians. The key idea of random projection arises
from the Johnson-Lindenstrauss lemma: if points in a vector space are
projected onto a randomly selected subspace of suitably high dimension,
then the distances between the points are approximately preserved.</p><p>Principal component analysis can be employed in a nonlinear way by means
of the kernel trick. The resulting technique is capable of constructing
nonlinear mappings that maximize the variance in the data. The resulting
technique is entitled Kernel PCA. Other prominent nonlinear techniques
include manifold learning techniques such as locally linear embedding
(LLE), Hessian LLE, Laplacian eigenmaps, and LTSA. These techniques
construct a low-dimensional data representation using a cost function
that retains local properties of the data, and can be viewed as defining
a graph-based kernel for Kernel PCA. More recently, techniques have been
proposed that, instead of defining a fixed kernel, try to learn the kernel
using semidefinite programming. The most prominent example of such a
technique is maximum variance unfolding (MVU). The central idea of MVU
is to exactly preserve all pairwise distances between nearest neighbors
(in the inner product space), while maximizing the distances between points
that are not nearest neighbors.</p><p>An alternative approach to neighborhood preservation is through the
minimization of a cost function that measures differences between
distances in the input and output spaces. Important examples of such
techniques include classical multidimensional scaling (which is identical
to PCA), Isomap (which uses geodesic distances in the data space), diffusion
maps (which uses diffusion distances in the data space), t-SNE (which
minimizes the divergence between distributions over pairs of points),
and curvilinear component analysis.</p><p>A different approach to nonlinear dimensionality reduction is through the
use of autoencoders, a special kind of feed-forward neural networks with
a bottle-neck hidden layer. The training of deep encoders is typically
performed using a greedy layer-wise pre-training (e.g., using a stack of
Restricted Boltzmann machines) that is followed by a finetuning stage based
on backpropagation.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li class="current-entities indented1">
                        <span class="separator"></span>
                        <a class="object" href="" title="Input operators."></a>
                        <a href="" title="Input operators.">
                          read
                        </a>
                      </li><li class="current-entities indented1">
                        <span class="separator"></span>
                        <a class="object" href="write$.html" title="Output operators."></a>
                        <a href="write$.html" title="Output operators.">
                          write
                        </a>
                      </li><li name="smile.regression" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="regression"></a><a id="regression:regression"></a>
      <span class="permalink">
      <a href="../smile/index.html#regression" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Regression analysis." href="regression/index.html">
              <span class="name">regression</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Regression analysis.</p><div class="fullcomment"><div class="comment cmt"><p>Regression analysis. Regression analysis includes any
techniques for modeling and analyzing several variables, when the focus
is on the relationship between a dependent variable and one or more
independent variables. Most commonly, regression analysis estimates the
conditional expectation of the dependent variable given the independent
variables. Therefore, the estimation target is a function of the independent
variables called the regression function. Regression analysis is widely
used for prediction and forecasting.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.sequence" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="sequence"></a><a id="sequence:sequence"></a>
      <span class="permalink">
      <a href="../smile/index.html#sequence" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Learning algorithms for sequence data." href="sequence/index.html">
              <span class="name">sequence</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Learning algorithms for sequence data.</p><div class="fullcomment"><div class="comment cmt"><p>Learning algorithms for sequence data.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.util" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="util"></a><a id="util:util"></a>
      <span class="permalink">
      <a href="../smile/index.html#util" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Utility functions." href="util/index.html">
              <span class="name">util</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Utility functions.</p><div class="fullcomment"><div class="comment cmt"><p>Utility functions.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.validation" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="validation"></a><a id="validation:validation"></a>
      <span class="permalink">
      <a href="../smile/index.html#validation" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Model validation." href="validation/index.html">
              <span class="name">validation</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Model validation.</p><div class="fullcomment"><div class="comment cmt"><p>Model validation.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.vq" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="vq"></a><a id="vq:vq"></a>
      <span class="permalink">
      <a href="../smile/index.html#vq" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="Originally used for data compression, Vector quantization (VQ)
allows the modeling of probability density functions by
the distribution of prototype vectors." href="vq/index.html">
              <span class="name">vq</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">Originally used for data compression, Vector quantization (VQ)
allows the modeling of probability density functions by
the distribution of prototype vectors.</p><div class="fullcomment"><div class="comment cmt"><p>Originally used for data compression, Vector quantization (VQ)
allows the modeling of probability density functions by
the distribution of prototype vectors. It works by dividing a large set of points
(vectors) into groups having approximately the same number of
points closest to them. Each group is represented by its centroid
point, as in K-Means and some other clustering algorithms.</p><p>Vector quantization is is based on the competitive learning paradigm,
and also closely related to sparse coding models
used in deep learning algorithms such as autoencoder.</p><p>Algorithms in this package also support the <code>partition</code>
method for clustering purpose.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li><li name="smile.wavelet" visbl="pub" class="indented2 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="wavelet"></a><a id="wavelet:wavelet"></a>
      <span class="permalink">
      <a href="../smile/index.html#wavelet" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">package</span>
      </span>
      <span class="symbol">
        <a title="A wavelet is a wave-like oscillation with an amplitude that starts out at
zero, increases, and then decreases back to zero." href="wavelet/index.html">
              <span class="name">wavelet</span>
            </a>
      </span>
      
      <p class="shortcomment cmt">A wavelet is a wave-like oscillation with an amplitude that starts out at
zero, increases, and then decreases back to zero.</p><div class="fullcomment"><div class="comment cmt"><p>A wavelet is a wave-like oscillation with an amplitude that starts out at
zero, increases, and then decreases back to zero. Like the fast Fourier
transform (FFT), the discrete wavelet transform (DWT) is a fast, linear
operation that operates on a data vector whose length is an integer power
of 2, transforming it into a numerically different vector of the same length.
The wavelet transform is invertible and in fact orthogonal. Both FFT and DWT
can be viewed as a rotation in function space.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="index.html" class="extype" name="smile">smile</a></dd></dl></div>
    </li>
              </ul>
            </div>
          </div>
          <div id="content">
            <body class="object value">
      <div id="definition">
        <div class="big-circle object">o</div>
        <p id="owner"><a href="index.html" class="extype" name="smile">smile</a></p>
        <h1>read<span class="permalink">
      <a href="../smile/read$.html" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span></h1>
        <h3><span class="morelinks"></span></h3>
      </div>

      <h4 id="signature" class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">object</span>
      </span>
      <span class="symbol">
        <span class="name">read</span>
      </span>
      </h4>

      
          <div id="comment" class="fullcommenttop"><div class="comment cmt"><p>Input operators.</p></div><div class="toggleContainer block">
          <span class="toggle">
            Linear Supertypes
          </span>
          <div class="superTypes hiddenContent"><span class="extype" name="scala.AnyRef">AnyRef</span>, <span class="extype" name="scala.Any">Any</span></div>
        </div></div>
        

      <div id="mbrsel">
        <div class="toggle"></div>
        <div id="memberfilter">
          <i class="material-icons arrow"></i>
          <span class="input">
            <input id="mbrsel-input" placeholder="Filter all members" type="text" accesskey="/" />
          </span>
          <i class="clear material-icons"></i>
        </div>
        <div id="filterby">
          <div id="order">
            <span class="filtertype">Ordering</span>
            <ol>
              
              <li class="alpha in"><span>Alphabetic</span></li>
              <li class="inherit out"><span>By Inheritance</span></li>
            </ol>
          </div>
          <div class="ancestors">
                  <span class="filtertype">Inherited<br />
                  </span>
                  <ol id="linearization">
                    <li class="in" name="smile.read"><span>read</span></li><li class="in" name="scala.AnyRef"><span>AnyRef</span></li><li class="in" name="scala.Any"><span>Any</span></li>
                  </ol>
                </div><div class="ancestors">
              <span class="filtertype"></span>
              <ol>
                <li class="hideall out"><span>Hide All</span></li>
                <li class="showall in"><span>Show All</span></li>
              </ol>
            </div>
          <div id="visbl">
              <span class="filtertype">Visibility</span>
              <ol><li class="public in"><span>Public</span></li><li class="all out"><span>All</span></li></ol>
            </div>
        </div>
      </div>

      <div id="template">
        <div id="allMembers">
        

        

        

        <div class="values members">
              <h3>Value Members</h3>
              <ol>
                <li name="scala.AnyRef#!=" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="!=(x$1:Any):Boolean"></a><a id="!=(Any):Boolean"></a>
      <span class="permalink">
      <a href="../smile/read$.html#!=(x$1:Any):Boolean" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span title="gt4s: $bang$eq" class="name">!=</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.Any">Any</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef → Any</dd></dl></div>
    </li><li name="scala.AnyRef###" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="##():Int"></a>
      <span class="permalink">
      <a href="../smile/read$.html###():Int" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span title="gt4s: $hash$hash" class="name">##</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Int">Int</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef → Any</dd></dl></div>
    </li><li name="scala.AnyRef#==" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="==(x$1:Any):Boolean"></a><a id="==(Any):Boolean"></a>
      <span class="permalink">
      <a href="../smile/read$.html#==(x$1:Any):Boolean" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span title="gt4s: $eq$eq" class="name">==</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.Any">Any</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef → Any</dd></dl></div>
    </li><li name="smile.read#apply" visbl="pub" class="indented0 " data-isabs="false" fullComment="no" group="Ungrouped">
      <a id="apply(file:String):AnyRef"></a><a id="apply(String):AnyRef"></a>
      <span class="permalink">
      <a href="../smile/read$.html#apply(file:String):AnyRef" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">apply</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="scala.AnyRef">AnyRef</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads a <code>Serializable</code> object/model.</p>
    </li><li name="smile.read#arff" visbl="pub" class="indented0 " data-isabs="false" fullComment="no" group="Ungrouped">
      <a id="arff(file:String,responseIndex:Int):smile.data.AttributeDataset"></a><a id="arff(String,Int):AttributeDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#arff(file:String,responseIndex:Int):smile.data.AttributeDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">arff</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>, <span name="responseIndex">responseIndex: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">1</span></span>)</span><span class="result">: <span class="extype" name="smile.data.AttributeDataset">AttributeDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads an ARFF file.</p>
    </li><li name="scala.Any#asInstanceOf" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="asInstanceOf[T0]:T0"></a>
      <span class="permalink">
      <a href="../smile/read$.html#asInstanceOf[T0]:T0" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">asInstanceOf</span><span class="tparams">[<span name="T0">T0</span>]</span><span class="result">: <span class="extype" name="scala.Any.asInstanceOf.T0">T0</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>Any</dd></dl></div>
    </li><li name="scala.AnyRef#clone" visbl="prt" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="clone():Object"></a><a id="clone():AnyRef"></a>
      <span class="permalink">
      <a href="../smile/read$.html#clone():Object" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">clone</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.AnyRef">AnyRef</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected[<a href="../java/lang/index.html" class="extype" name="java.lang">java.lang</a>] </dd><dt>Definition Classes</dt><dd>AnyRef</dd><dt>Annotations</dt><dd>
                <span class="name">@throws</span><span class="args">(<span>
      
      <span class="defval" name="classOf[java.lang.CloneNotSupportedException]">...</span>
    </span>)</span>
              
        </dd></dl></div>
    </li><li name="smile.read#coo" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="coo(file:String,arrayIndexOrigin:Int):smile.data.SparseDataset"></a><a id="coo(String,Int):SparseDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#coo(file:String,arrayIndexOrigin:Int):smile.data.SparseDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">coo</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>, <span name="arrayIndexOrigin">arrayIndexOrigin: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">0</span></span>)</span><span class="result">: <span class="extype" name="smile.data.SparseDataset">SparseDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads spare dataset in coordinate triple tuple list format.</p><div class="fullcomment"><div class="comment cmt"><p>Reads spare dataset in coordinate triple tuple list format.
Coordinate file stores a list of (row, column, value) tuples:</p><pre>instanceID attributeID value
instanceID attributeID value
instanceID attributeID value
instanceID attributeID value
...
instanceID attributeID value
instanceID attributeID value
instanceID attributeID value</pre><p>Ideally, the entries are sorted (by row index, then column index) to
improve random access times. This format is good for incremental matrix
construction.</p><p>Optionally, there may be 2 header lines</p><pre>D    <span class="cmt">// The number of instances</span>
W    <span class="cmt">// The number of attributes</span></pre><p>or 3 header lines</p><pre>D    <span class="cmt">// The number of instances</span>
W    <span class="cmt">// The number of attributes</span>
N    <span class="cmt">// The total number of nonzero items in the dataset.</span></pre><p>These header lines will be ignored.
</p></div><dl class="paramcmts block"><dt class="param">file</dt><dd class="cmt"><p>the file path.</p></dd><dt class="param">arrayIndexOrigin</dt><dd class="cmt"><p>the starting index of array. By default, it is
0 as in C/C++ and Java. But it could be 1 to parse data produced
by other programming language such as Fortran.</p></dd></dl></div>
    </li><li name="smile.read#csv" visbl="pub" class="indented0 " data-isabs="false" fullComment="no" group="Ungrouped">
      <a id="csv(file:String,response:Option[(smile.data.Attribute,Int)],comment:String,missing:String,header:Boolean,rowNames:Boolean):smile.data.AttributeDataset"></a><a id="csv(String,Option[(Attribute,Int)],String,String,Boolean,Boolean):AttributeDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#csv(file:String,response:Option[(smile.data.Attribute,Int)],comment:String,missing:String,header:Boolean,rowNames:Boolean):smile.data.AttributeDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">csv</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>, <span name="response">response: <span class="extype" name="scala.Option">Option</span>[(<span class="extype" name="smile.data.Attribute">Attribute</span>, <span class="extype" name="scala.Int">Int</span>)] = <span class="symbol">None</span></span>, <span name="comment">comment: <span class="extype" name="scala.Predef.String">String</span> = <span class="symbol">&quot;%&quot;</span></span>, <span name="missing">missing: <span class="extype" name="scala.Predef.String">String</span> = <span class="symbol">&quot;?&quot;</span></span>, <span name="header">header: <span class="extype" name="scala.Boolean">Boolean</span> = <span class="symbol">false</span></span>, <span name="rowNames">rowNames: <span class="extype" name="scala.Boolean">Boolean</span> = <span class="symbol">false</span></span>)</span><span class="result">: <span class="extype" name="smile.data.AttributeDataset">AttributeDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads a CSV file  with response variable.</p>
    </li><li name="scala.AnyRef#eq" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="eq(x$1:AnyRef):Boolean"></a><a id="eq(AnyRef):Boolean"></a>
      <span class="permalink">
      <a href="../smile/read$.html#eq(x$1:AnyRef):Boolean" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">eq</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.AnyRef">AnyRef</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="scala.AnyRef#equals" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="equals(x$1:Any):Boolean"></a><a id="equals(Any):Boolean"></a>
      <span class="permalink">
      <a href="../smile/read$.html#equals(x$1:Any):Boolean" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">equals</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.Any">Any</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef → Any</dd></dl></div>
    </li><li name="scala.AnyRef#finalize" visbl="prt" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="finalize():Unit"></a>
      <span class="permalink">
      <a href="../smile/read$.html#finalize():Unit" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">finalize</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected[<a href="../java/lang/index.html" class="extype" name="java.lang">java.lang</a>] </dd><dt>Definition Classes</dt><dd>AnyRef</dd><dt>Annotations</dt><dd>
                <span class="name">@throws</span><span class="args">(<span>
      
      <span class="symbol">classOf[java.lang.Throwable]</span>
    </span>)</span>
              
        </dd></dl></div>
    </li><li name="smile.read#gct" visbl="pub" class="indented0 " data-isabs="false" fullComment="no" group="Ungrouped">
      <a id="gct(file:String):smile.data.AttributeDataset"></a><a id="gct(String):AttributeDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#gct(file:String):smile.data.AttributeDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">gct</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="smile.data.AttributeDataset">AttributeDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads GCT microarray gene expression file.</p>
    </li><li name="scala.AnyRef#getClass" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="getClass():Class[_]"></a>
      <span class="permalink">
      <a href="../smile/read$.html#getClass():Class[_]" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">getClass</span><span class="params">()</span><span class="result">: <span class="extype" name="java.lang.Class">Class</span>[_]</span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef → Any</dd></dl></div>
    </li><li name="scala.AnyRef#hashCode" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="hashCode():Int"></a>
      <span class="permalink">
      <a href="../smile/read$.html#hashCode():Int" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">hashCode</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Int">Int</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef → Any</dd></dl></div>
    </li><li name="smile.read#hb" visbl="pub" class="indented0 " data-isabs="false" fullComment="no" group="Ungrouped">
      <a id="hb(file:String):smile.math.matrix.SparseMatrix"></a><a id="hb(String):SparseMatrix"></a>
      <span class="permalink">
      <a href="../smile/read$.html#hb(file:String):smile.math.matrix.SparseMatrix" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">hb</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="smile.math.matrix.SparseMatrix">SparseMatrix</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads Harwell-Boeing column-compressed sparse matrix.</p>
    </li><li name="scala.Any#isInstanceOf" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="isInstanceOf[T0]:Boolean"></a>
      <span class="permalink">
      <a href="../smile/read$.html#isInstanceOf[T0]:Boolean" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">isInstanceOf</span><span class="tparams">[<span name="T0">T0</span>]</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>Any</dd></dl></div>
    </li><li name="smile.read#jdbc" visbl="pub" class="indented0 " data-isabs="false" fullComment="no" group="Ungrouped">
      <a id="jdbc(rs:java.sql.ResultSet):smile.data.AttributeDataset"></a><a id="jdbc(ResultSet):AttributeDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#jdbc(rs:java.sql.ResultSet):smile.data.AttributeDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">jdbc</span><span class="params">(<span name="rs">rs: <span class="extype" name="java.sql.ResultSet">ResultSet</span></span>)</span><span class="result">: <span class="extype" name="smile.data.AttributeDataset">AttributeDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads a JDBC query result to an AttributeDataset.</p>
    </li><li name="smile.read#libsvm" visbl="pub" class="indented0 " data-isabs="false" fullComment="no" group="Ungrouped">
      <a id="libsvm(file:String):smile.data.SparseDataset"></a><a id="libsvm(String):SparseDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#libsvm(file:String):smile.data.SparseDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">libsvm</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="smile.data.SparseDataset">SparseDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads a LivSVM file.</p>
    </li><li name="scala.AnyRef#ne" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="ne(x$1:AnyRef):Boolean"></a><a id="ne(AnyRef):Boolean"></a>
      <span class="permalink">
      <a href="../smile/read$.html#ne(x$1:AnyRef):Boolean" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">ne</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.AnyRef">AnyRef</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="scala.AnyRef#notify" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="notify():Unit"></a>
      <span class="permalink">
      <a href="../smile/read$.html#notify():Unit" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">notify</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="scala.AnyRef#notifyAll" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="notifyAll():Unit"></a>
      <span class="permalink">
      <a href="../smile/read$.html#notifyAll():Unit" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">notifyAll</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="smile.read#pcl" visbl="pub" class="indented0 " data-isabs="false" fullComment="no" group="Ungrouped">
      <a id="pcl(file:String):smile.data.AttributeDataset"></a><a id="pcl(String):AttributeDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#pcl(file:String):smile.data.AttributeDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">pcl</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="smile.data.AttributeDataset">AttributeDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads PCL microarray gene expression file.</p>
    </li><li name="smile.read#res" visbl="pub" class="indented0 " data-isabs="false" fullComment="no" group="Ungrouped">
      <a id="res(file:String):smile.data.AttributeDataset"></a><a id="res(String):AttributeDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#res(file:String):smile.data.AttributeDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">res</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="smile.data.AttributeDataset">AttributeDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads RES microarray gene expression file.</p>
    </li><li name="smile.read#sb" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="sb(file:String):smile.data.BinarySparseDataset"></a><a id="sb(String):BinarySparseDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#sb(file:String):smile.data.BinarySparseDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">sb</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="smile.data.BinarySparseDataset">BinarySparseDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads sparse binary dataset.</p><div class="fullcomment"><div class="comment cmt"><p>Reads sparse binary dataset. Each item is stored as an integer array, which
are the indices of nonzero elements in ascending order.
</p></div></div>
    </li><li name="scala.AnyRef#synchronized" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="synchronized[T0](x$1:=&gt;T0):T0"></a><a id="synchronized[T0](⇒T0):T0"></a>
      <span class="permalink">
      <a href="../smile/read$.html#synchronized[T0](x$1:=&gt;T0):T0" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">synchronized</span><span class="tparams">[<span name="T0">T0</span>]</span><span class="params">(<span name="arg0">arg0: ⇒ <span class="extype" name="java.lang.AnyRef.synchronized.T0">T0</span></span>)</span><span class="result">: <span class="extype" name="java.lang.AnyRef.synchronized.T0">T0</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="smile.read#table" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="table(file:String,response:Option[(smile.data.Attribute,Int)],delimiter:String,comment:String,missing:String,header:Boolean,rowNames:Boolean):smile.data.AttributeDataset"></a><a id="table(String,Option[(Attribute,Int)],String,String,String,Boolean,Boolean):AttributeDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#table(file:String,response:Option[(smile.data.Attribute,Int)],delimiter:String,comment:String,missing:String,header:Boolean,rowNames:Boolean):smile.data.AttributeDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">table</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>, <span name="response">response: <span class="extype" name="scala.Option">Option</span>[(<span class="extype" name="smile.data.Attribute">Attribute</span>, <span class="extype" name="scala.Int">Int</span>)] = <span class="symbol">None</span></span>, <span name="delimiter">delimiter: <span class="extype" name="scala.Predef.String">String</span> = <span class="symbol">&quot;\\s+&quot;</span></span>, <span name="comment">comment: <span class="extype" name="scala.Predef.String">String</span> = <span class="symbol">&quot;%&quot;</span></span>, <span name="missing">missing: <span class="extype" name="scala.Predef.String">String</span> = <span class="symbol">&quot;?&quot;</span></span>, <span name="header">header: <span class="extype" name="scala.Boolean">Boolean</span> = <span class="symbol">false</span></span>, <span name="rowNames">rowNames: <span class="extype" name="scala.Boolean">Boolean</span> = <span class="symbol">false</span></span>)</span><span class="result">: <span class="extype" name="smile.data.AttributeDataset">AttributeDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads a delimited text file with response variable.</p><div class="fullcomment"><div class="comment cmt"><p>Reads a delimited text file with response variable. By default, the parser expects a
white-space-separated-values file. Each line in the file corresponds
to a row in the table. Within a line, fields are separated by white spaces,
each field belonging to one table column. This class can also be
used to read other text tabular files by setting delimiter character
such ash ','. The file may contain comment lines (starting with '%')
and missing values (indicated by placeholder '?').
</p></div><dl class="paramcmts block"><dt class="param">file</dt><dd class="cmt"><p>the file path</p></dd><dt class="param">response</dt><dd class="cmt"><p>optional response variable attribute and the column index of response variable. The column index starts at 0.</p></dd><dt class="param">delimiter</dt><dd class="cmt"><p>delimiter string</p></dd><dt class="param">comment</dt><dd class="cmt"><p>the start of comment lines</p></dd><dt class="param">missing</dt><dd class="cmt"><p>the missing value placeholder</p></dd><dt class="param">header</dt><dd class="cmt"><p>true if the first row is header/column names</p></dd><dt class="param">rowNames</dt><dd class="cmt"><p>true if the first column is row id/names</p></dd><dt>returns</dt><dd class="cmt"><p>an attribute dataset</p></dd></dl></div>
    </li><li name="scala.AnyRef#toString" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="toString():String"></a>
      <span class="permalink">
      <a href="../smile/read$.html#toString():String" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">toString</span><span class="params">()</span><span class="result">: <span class="extype" name="java.lang.String">String</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef → Any</dd></dl></div>
    </li><li name="smile.read#txt" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="txt(file:String):smile.data.AttributeDataset"></a><a id="txt(String):AttributeDataset"></a>
      <span class="permalink">
      <a href="../smile/read$.html#txt(file:String):smile.data.AttributeDataset" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">txt</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="smile.data.AttributeDataset">AttributeDataset</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads TXT microarray gene expression file.</p><div class="fullcomment"><div class="comment cmt"><p>Reads TXT microarray gene expression file.
The TXT format is a tab delimited file
format that describes an expression dataset. It is organized as follows:</p><p>The first line contains the labels Name and Description followed by the
identifiers for each sample in the dataset. The Description is optional.</p><p>Line format:</p><pre>Name(tab)Description(tab)(sample <span class="num">1</span> name)(tab)(sample <span class="num">2</span> name) (tab) ... (sample N name)</pre><p>Example:</p><pre>Name Description DLBC1_1 DLBC2_1 ... DLBC58_0</pre><p>The remainder of the file contains data for each of the genes. There is one
line for each gene. Each line contains the gene name, gene description, and
a value for each sample in the dataset. If the first line contains the
Description label, include a description for each gene. If the first line
does not contain the Description label, do not include descriptions for
any gene. Gene names and descriptions can contain spaces since fields are
separated by tabs.</p><p>Line format:</p><pre>(gene name) (tab) (gene description) (tab) (col <span class="num">1</span> data) (tab) (col <span class="num">2</span> data) (tab) ... (col N data)</pre><p>Example:</p><pre>AFFX-BioB-<span class="num">5</span>_at AFFX-BioB-<span class="num">5</span>_at (endogenous control) -<span class="num">104</span> -<span class="num">152</span> -<span class="num">158</span> ... -<span class="num">44</span></pre></div></div>
    </li><li name="scala.AnyRef#wait" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="wait():Unit"></a>
      <span class="permalink">
      <a href="../smile/read$.html#wait():Unit" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">wait</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd><dt>Annotations</dt><dd>
                <span class="name">@throws</span><span class="args">(<span>
      
      <span class="defval" name="classOf[java.lang.InterruptedException]">...</span>
    </span>)</span>
              
        </dd></dl></div>
    </li><li name="scala.AnyRef#wait" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="wait(x$1:Long,x$2:Int):Unit"></a><a id="wait(Long,Int):Unit"></a>
      <span class="permalink">
      <a href="../smile/read$.html#wait(x$1:Long,x$2:Int):Unit" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">wait</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.Long">Long</span></span>, <span name="arg1">arg1: <span class="extype" name="scala.Int">Int</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd><dt>Annotations</dt><dd>
                <span class="name">@throws</span><span class="args">(<span>
      
      <span class="defval" name="classOf[java.lang.InterruptedException]">...</span>
    </span>)</span>
              
        </dd></dl></div>
    </li><li name="scala.AnyRef#wait" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="wait(x$1:Long):Unit"></a><a id="wait(Long):Unit"></a>
      <span class="permalink">
      <a href="../smile/read$.html#wait(x$1:Long):Unit" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">wait</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.Long">Long</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd><dt>Annotations</dt><dd>
                <span class="name">@throws</span><span class="args">(<span>
      
      <span class="defval" name="classOf[java.lang.InterruptedException]">...</span>
    </span>)</span>
              
        </dd></dl></div>
    </li><li name="smile.read#wavefront" visbl="pub" class="indented0 " data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="wavefront(file:String):(Array[Array[Double]],Array[Array[Int]])"></a><a id="wavefront(String):(Array[Array[Double]],Array[Array[Int]])"></a>
      <span class="permalink">
      <a href="../smile/read$.html#wavefront(file:String):(Array[Array[Double]],Array[Array[Int]])" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">wavefront</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: (<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Double">Double</span>]], <span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Array">Array</span>[<span class="extype" name="scala.Int">Int</span>]])</span>
      </span>
      
      <p class="shortcomment cmt">Reads a Wavefront OBJ file.</p><div class="fullcomment"><div class="comment cmt"><p>Reads a Wavefront OBJ file. The OBJ file format is a simple format of 3D geometry including
the position of each vertex, the UV position of each texture coordinate vertex,
vertex normals, and the faces that make each polygon defined as a list of vertices,
and texture vertices. Vertices are stored in a counter-clockwise order by default,
making explicit declaration of face normals unnecessary. OBJ coordinates have no units,
but OBJ files can contain scale information in a human readable comment line.</p><p>Note that we parse only vertex and face elements. All other information ignored.
</p></div><dl class="paramcmts block"><dt class="param">file</dt><dd class="cmt"><p>the file path</p></dd><dt>returns</dt><dd class="cmt"><p>a tuple of vertex array and edge array.</p></dd></dl></div>
    </li><li name="smile.read#xstream" visbl="pub" class="indented0 " data-isabs="false" fullComment="no" group="Ungrouped">
      <a id="xstream(file:String):AnyRef"></a><a id="xstream(String):AnyRef"></a>
      <span class="permalink">
      <a href="../smile/read$.html#xstream(file:String):AnyRef" title="Permalink">
        <i class="material-icons"></i>
      </a>
    </span>
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">xstream</span><span class="params">(<span name="file">file: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="scala.AnyRef">AnyRef</span></span>
      </span>
      
      <p class="shortcomment cmt">Reads an object/model that was serialized by XStream.</p>
    </li>
              </ol>
            </div>

        

        
        </div>

        <div id="inheritedMembers">
        <div class="parent" name="scala.AnyRef">
              <h3>Inherited from <span class="extype" name="scala.AnyRef">AnyRef</span></h3>
            </div><div class="parent" name="scala.Any">
              <h3>Inherited from <span class="extype" name="scala.Any">Any</span></h3>
            </div>
        
        </div>

        <div id="groupedMembers">
        <div class="group" name="Ungrouped">
              <h3>Ungrouped</h3>
              
            </div>
        </div>

      </div>

      <div id="tooltip"></div>

      <div id="footer">  </div>
    </body>
          </div>
        </div>
      </div>
    </body>
      </html>
